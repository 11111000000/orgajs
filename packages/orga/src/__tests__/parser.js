const Parser = require('../parser')

describe('Parser', () => {
  let parser = new Parser()
  it('works', () => {
    const content = `
#+TITLE: hello world
#+TODO: TODO NEXT | DONE
#+DATE: 2018-01-01

* NEXT headline one
DEADLINE: <2018-01-01 Mon>
:PROPERTIES:
key: value
key: value
:END:

[[https://github.com/xiaoxinghu/orgajs][Here's]] to the *crazy* ones, the /misfits/, the _rebels_, the ~troublemakers~,
the round pegs in the +round+ square holes...
`
    const document = parser.parse(content)
    expect(document).toMatchSnapshot()
  })

  it(`can handle blocks`, () => {
    const content = `
* headline
#+BEGIN_SRC javascript
console.log('hello')
console.log('world')
#+END_SRC
`
    expect(parser.parse(content)).toMatchSnapshot()
  })

  it(`can handle broken blocks`, () => {
    const endLess = `
* headline
#+BEGIN_SRC javascript
console.log('hello')
console.log('world')
`
    expect(parser.parse(endLess)).toMatchSnapshot()

    const beginLess = `
* headline
console.log('hello')
console.log('world')
#+END_SRC
`
    expect(parser.parse(beginLess)).toMatchSnapshot()
  })

  it('can handle drawers', () => {
    const content = `
* headline
:PROPERTIES:
key1: value 1
key2: value 2
:END:
`
    expect(parser.parse(content)).toMatchSnapshot()
  })

  it('can handle broken drawers', () => {
    const content = `
* headline
:PROPERTIES:
key: value
key: value

Paragraph
`
    expect(parser.parse(content)).toMatchSnapshot()
  })

  it('can handle nested headlines', () => {
    const content = `
* #HEADLINE# 1
Paragraph
** #HEADLINE# 1.1
*** #HEADLINE# 1.1.1
content

** #HEADLINE# 1.2
* #HEADLINE# 2
** #HEADLINE# 2.2
`
    expect(parser.parse(content)).toMatchSnapshot()

  })

  it('can handle unordered list', () => {
    const content = `
- apple
- banana
- orange
`
    expect(parser.parse(content)).toMatchSnapshot()
  })

  it('can handle ordered list', () => {
    const content = `
1. apple
5. banana
- orange
`
    expect(parser.parse(content)).toMatchSnapshot()
  })

  it('can handle nested list', () => {
    const content = `
1. apple
  - iPhone
  - Mac
5. banana
- orange
`
    expect(parser.parse(content)).toMatchSnapshot()
  })

  it('can handle table', () => {
    const content = `
| Name         | Species    | Gender | Role         |
|--------------+------------+--------+--------------|
| Bruce Wayne  | Human      | M      | Batman       |
| Clark Kent   | Kryptonian | M      | Superman     |
| Diana Prince | Amazonian  | F      | Wonder Woman |
`
    expect(parser.parse(content)).toMatchSnapshot()
  })

  it('can handle html export', () => {
    const content = `#+HTML: <h1>Hello</h1>`
    expect(parser.parse(content)).toMatchSnapshot()
  })
  it('can handle html export block', () => {
    const content = `
#+BEGIN_EXPORT html
<h1>hello</h1>
<p>world!</p>
#+END_EXPORT
`
    expect(parser.parse(content)).toMatchSnapshot()
  })
})
