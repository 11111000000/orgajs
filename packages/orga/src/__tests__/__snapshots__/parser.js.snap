// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parser can handle affiliated keywords 1`] = `
Object {
  "children": Array [
    Object {
      "attributes": Object {
        "NAME": "code_block",
      },
      "children": Array [],
      "name": "SRC",
      "params": Array [
        "javascript",
      ],
      "parent": [Circular],
      "type": "block",
      "value": "console.log('hello world.')",
    },
  ],
  "meta": Object {},
  "type": "root",
}
`;

exports[`Parser can handle blocks 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "parent": [Circular],
              "type": "text",
              "value": "headline",
            },
          ],
          "keyword": undefined,
          "level": 1,
          "parent": [Circular],
          "priority": undefined,
          "tags": Array [],
          "type": "headline",
        },
        Object {
          "children": Array [],
          "name": "SRC",
          "params": Array [
            "javascript",
          ],
          "parent": [Circular],
          "type": "block",
          "value": "console.log('hello')
console.log('world')",
        },
      ],
      "level": 1,
      "parent": [Circular],
      "type": "section",
    },
  ],
  "meta": Object {},
  "type": "root",
}
`;

exports[`Parser can handle broken blocks 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "parent": [Circular],
              "type": "text",
              "value": "headline",
            },
          ],
          "keyword": undefined,
          "level": 1,
          "parent": [Circular],
          "priority": undefined,
          "tags": Array [],
          "type": "headline",
        },
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "parent": [Circular],
              "type": "text",
              "value": "#+BEGIN_SRC javascript",
            },
            Object {
              "children": Array [],
              "parent": [Circular],
              "type": "text",
              "value": "console.log('hello')",
            },
            Object {
              "children": Array [],
              "parent": [Circular],
              "type": "text",
              "value": "console.log('world')",
            },
          ],
          "parent": [Circular],
          "type": "paragraph",
        },
      ],
      "level": 1,
      "parent": [Circular],
      "type": "section",
    },
  ],
  "meta": Object {},
  "type": "root",
}
`;

exports[`Parser can handle broken blocks 2`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "parent": [Circular],
              "type": "text",
              "value": "headline",
            },
          ],
          "keyword": undefined,
          "level": 1,
          "parent": [Circular],
          "priority": undefined,
          "tags": Array [],
          "type": "headline",
        },
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "parent": [Circular],
              "type": "text",
              "value": "console.log('hello')",
            },
            Object {
              "children": Array [],
              "parent": [Circular],
              "type": "text",
              "value": "console.log('world')",
            },
            Object {
              "children": Array [],
              "parent": [Circular],
              "type": "text",
              "value": "#+END_SRC",
            },
          ],
          "parent": [Circular],
          "type": "paragraph",
        },
      ],
      "level": 1,
      "parent": [Circular],
      "type": "section",
    },
  ],
  "meta": Object {},
  "type": "root",
}
`;

exports[`Parser can handle broken drawers 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "parent": [Circular],
              "type": "text",
              "value": "headline",
            },
          ],
          "keyword": undefined,
          "level": 1,
          "parent": [Circular],
          "priority": undefined,
          "tags": Array [],
          "type": "headline",
        },
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "parent": [Circular],
              "type": "text",
              "value": ":PROPERTIES:",
            },
            Object {
              "children": Array [],
              "parent": [Circular],
              "type": "text",
              "value": "key: value",
            },
            Object {
              "children": Array [],
              "parent": [Circular],
              "type": "text",
              "value": "key: value",
            },
          ],
          "parent": [Circular],
          "type": "paragraph",
        },
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "parent": [Circular],
              "type": "text",
              "value": "Paragraph",
            },
          ],
          "parent": [Circular],
          "type": "paragraph",
        },
      ],
      "level": 1,
      "parent": [Circular],
      "type": "section",
    },
  ],
  "meta": Object {},
  "type": "root",
}
`;

exports[`Parser can handle drawers 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "parent": [Circular],
              "type": "text",
              "value": "headline",
            },
            Object {
              "children": Array [],
              "name": "PROPERTIES",
              "parent": [Circular],
              "type": "drawer",
              "value": "key1: value 1
key2: value 2",
            },
          ],
          "keyword": undefined,
          "level": 1,
          "parent": [Circular],
          "priority": undefined,
          "tags": Array [],
          "type": "headline",
        },
      ],
      "level": 1,
      "parent": [Circular],
      "type": "section",
    },
  ],
  "meta": Object {},
  "type": "root",
}
`;

exports[`Parser can handle html export 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [],
      "parent": [Circular],
      "type": "html",
      "value": "<h1>Hello</h1>",
    },
  ],
  "meta": Object {},
  "type": "root",
}
`;

exports[`Parser can handle html export block 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [],
      "parent": [Circular],
      "type": "html",
      "value": "<h1>hello</h1>
<p>world!</p>",
    },
  ],
  "meta": Object {},
  "type": "root",
}
`;

exports[`Parser can handle invalid affiliated keywords 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [],
      "name": "SRC",
      "params": Array [
        "javascript",
      ],
      "parent": [Circular],
      "type": "block",
      "value": "console.log('hello world.')",
    },
  ],
  "meta": Object {},
  "type": "root",
}
`;

exports[`Parser can handle invalid affiliated keywords 2`] = `
Object {
  "children": Array [
    Object {
      "children": Array [],
      "name": "SRC",
      "params": Array [
        "javascript",
      ],
      "parent": [Circular],
      "type": "block",
      "value": "console.log('hello world.')",
    },
  ],
  "meta": Object {
    "nop": "code_block",
  },
  "type": "root",
}
`;

exports[`Parser can handle nested headlines 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "parent": [Circular],
              "type": "text",
              "value": "#HEADLINE# 1",
            },
          ],
          "keyword": undefined,
          "level": 1,
          "parent": [Circular],
          "priority": undefined,
          "tags": Array [],
          "type": "headline",
        },
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "parent": [Circular],
              "type": "text",
              "value": "Paragraph",
            },
          ],
          "parent": [Circular],
          "type": "paragraph",
        },
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "children": Array [],
                  "parent": [Circular],
                  "type": "text",
                  "value": "#HEADLINE# 1.1",
                },
              ],
              "keyword": undefined,
              "level": 2,
              "parent": [Circular],
              "priority": undefined,
              "tags": Array [],
              "type": "headline",
            },
            Object {
              "children": Array [
                Object {
                  "children": Array [
                    Object {
                      "children": Array [],
                      "parent": [Circular],
                      "type": "text",
                      "value": "#HEADLINE# 1.1.1",
                    },
                  ],
                  "keyword": undefined,
                  "level": 3,
                  "parent": [Circular],
                  "priority": undefined,
                  "tags": Array [],
                  "type": "headline",
                },
                Object {
                  "children": Array [
                    Object {
                      "children": Array [],
                      "parent": [Circular],
                      "type": "text",
                      "value": "content",
                    },
                  ],
                  "parent": [Circular],
                  "type": "paragraph",
                },
              ],
              "level": 3,
              "parent": [Circular],
              "type": "section",
            },
          ],
          "level": 2,
          "parent": [Circular],
          "type": "section",
        },
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "children": Array [],
                  "parent": [Circular],
                  "type": "text",
                  "value": "#HEADLINE# 1.2",
                },
              ],
              "keyword": undefined,
              "level": 2,
              "parent": [Circular],
              "priority": undefined,
              "tags": Array [],
              "type": "headline",
            },
          ],
          "level": 2,
          "parent": [Circular],
          "type": "section",
        },
      ],
      "level": 1,
      "parent": [Circular],
      "type": "section",
    },
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "parent": [Circular],
              "type": "text",
              "value": "#HEADLINE# 2",
            },
          ],
          "keyword": undefined,
          "level": 1,
          "parent": [Circular],
          "priority": undefined,
          "tags": Array [],
          "type": "headline",
        },
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "children": Array [],
                  "parent": [Circular],
                  "type": "text",
                  "value": "#HEADLINE# 2.2",
                },
              ],
              "keyword": undefined,
              "level": 2,
              "parent": [Circular],
              "priority": undefined,
              "tags": Array [],
              "type": "headline",
            },
          ],
          "level": 2,
          "parent": [Circular],
          "type": "section",
        },
      ],
      "level": 1,
      "parent": [Circular],
      "type": "section",
    },
  ],
  "meta": Object {},
  "type": "root",
}
`;

exports[`Parser can handle nested list 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "parent": [Circular],
              "type": "text",
              "value": "apple",
            },
            Object {
              "children": Array [
                Object {
                  "children": Array [
                    Object {
                      "children": Array [],
                      "parent": [Circular],
                      "type": "text",
                      "value": "iPhone",
                    },
                  ],
                  "ordered": false,
                  "parent": [Circular],
                  "type": "listItem",
                },
                Object {
                  "children": Array [
                    Object {
                      "children": Array [],
                      "parent": [Circular],
                      "type": "text",
                      "value": "Mac",
                    },
                  ],
                  "ordered": false,
                  "parent": [Circular],
                  "type": "listItem",
                },
              ],
              "ordered": false,
              "parent": [Circular],
              "type": "list",
            },
          ],
          "ordered": true,
          "parent": [Circular],
          "type": "listItem",
        },
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "parent": [Circular],
              "type": "text",
              "value": "banana",
            },
          ],
          "ordered": true,
          "parent": [Circular],
          "type": "listItem",
        },
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "parent": [Circular],
              "type": "text",
              "value": "orange",
            },
          ],
          "ordered": false,
          "parent": [Circular],
          "type": "listItem",
        },
      ],
      "ordered": true,
      "parent": [Circular],
      "type": "list",
    },
  ],
  "meta": Object {},
  "type": "root",
}
`;

exports[`Parser can handle ordered list 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "parent": [Circular],
              "type": "text",
              "value": "apple",
            },
          ],
          "ordered": true,
          "parent": [Circular],
          "type": "listItem",
        },
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "parent": [Circular],
              "type": "text",
              "value": "banana",
            },
          ],
          "ordered": true,
          "parent": [Circular],
          "type": "listItem",
        },
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "parent": [Circular],
              "type": "text",
              "value": "orange",
            },
          ],
          "ordered": false,
          "parent": [Circular],
          "type": "listItem",
        },
      ],
      "ordered": true,
      "parent": [Circular],
      "type": "list",
    },
  ],
  "meta": Object {},
  "type": "root",
}
`;

exports[`Parser can handle table 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "children": Array [],
                  "parent": [Circular],
                  "type": "text",
                  "value": "Name",
                },
              ],
              "parent": [Circular],
              "type": "tableCell",
            },
            Object {
              "children": Array [
                Object {
                  "children": Array [],
                  "parent": [Circular],
                  "type": "text",
                  "value": "Species",
                },
              ],
              "parent": [Circular],
              "type": "tableCell",
            },
            Object {
              "children": Array [
                Object {
                  "children": Array [],
                  "parent": [Circular],
                  "type": "text",
                  "value": "Gender",
                },
              ],
              "parent": [Circular],
              "type": "tableCell",
            },
            Object {
              "children": Array [
                Object {
                  "children": Array [],
                  "parent": [Circular],
                  "type": "text",
                  "value": "Role",
                },
              ],
              "parent": [Circular],
              "type": "tableCell",
            },
          ],
          "parent": [Circular],
          "type": "tableRow",
        },
        Object {
          "children": Array [],
          "parent": [Circular],
          "type": "tableSeparator",
        },
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "children": Array [],
                  "parent": [Circular],
                  "type": "text",
                  "value": "Bruce Wayne",
                },
              ],
              "parent": [Circular],
              "type": "tableCell",
            },
            Object {
              "children": Array [
                Object {
                  "children": Array [],
                  "parent": [Circular],
                  "type": "text",
                  "value": "Human",
                },
              ],
              "parent": [Circular],
              "type": "tableCell",
            },
            Object {
              "children": Array [
                Object {
                  "children": Array [],
                  "parent": [Circular],
                  "type": "text",
                  "value": "M",
                },
              ],
              "parent": [Circular],
              "type": "tableCell",
            },
            Object {
              "children": Array [
                Object {
                  "children": Array [],
                  "parent": [Circular],
                  "type": "text",
                  "value": "Batman",
                },
              ],
              "parent": [Circular],
              "type": "tableCell",
            },
          ],
          "parent": [Circular],
          "type": "tableRow",
        },
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "children": Array [],
                  "parent": [Circular],
                  "type": "text",
                  "value": "Clark Kent",
                },
              ],
              "parent": [Circular],
              "type": "tableCell",
            },
            Object {
              "children": Array [
                Object {
                  "children": Array [],
                  "parent": [Circular],
                  "type": "text",
                  "value": "Kryptonian",
                },
              ],
              "parent": [Circular],
              "type": "tableCell",
            },
            Object {
              "children": Array [
                Object {
                  "children": Array [],
                  "parent": [Circular],
                  "type": "text",
                  "value": "M",
                },
              ],
              "parent": [Circular],
              "type": "tableCell",
            },
            Object {
              "children": Array [
                Object {
                  "children": Array [],
                  "parent": [Circular],
                  "type": "text",
                  "value": "Superman",
                },
              ],
              "parent": [Circular],
              "type": "tableCell",
            },
          ],
          "parent": [Circular],
          "type": "tableRow",
        },
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "children": Array [],
                  "parent": [Circular],
                  "type": "text",
                  "value": "Diana Prince",
                },
              ],
              "parent": [Circular],
              "type": "tableCell",
            },
            Object {
              "children": Array [
                Object {
                  "children": Array [],
                  "parent": [Circular],
                  "type": "text",
                  "value": "Amazonian",
                },
              ],
              "parent": [Circular],
              "type": "tableCell",
            },
            Object {
              "children": Array [
                Object {
                  "children": Array [],
                  "parent": [Circular],
                  "type": "text",
                  "value": "F",
                },
              ],
              "parent": [Circular],
              "type": "tableCell",
            },
            Object {
              "children": Array [
                Object {
                  "children": Array [],
                  "parent": [Circular],
                  "type": "text",
                  "value": "Wonder Woman",
                },
              ],
              "parent": [Circular],
              "type": "tableCell",
            },
          ],
          "parent": [Circular],
          "type": "tableRow",
        },
      ],
      "parent": [Circular],
      "type": "table",
    },
  ],
  "meta": Object {},
  "type": "root",
}
`;

exports[`Parser can handle unordered list 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "parent": [Circular],
              "type": "text",
              "value": "apple",
            },
          ],
          "ordered": false,
          "parent": [Circular],
          "type": "listItem",
        },
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "parent": [Circular],
              "type": "text",
              "value": "banana",
            },
          ],
          "ordered": false,
          "parent": [Circular],
          "type": "listItem",
        },
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "parent": [Circular],
              "type": "text",
              "value": "orange",
            },
          ],
          "ordered": false,
          "parent": [Circular],
          "type": "listItem",
        },
      ],
      "ordered": false,
      "parent": [Circular],
      "type": "list",
    },
  ],
  "meta": Object {},
  "type": "root",
}
`;

exports[`Parser works 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "parent": [Circular],
              "type": "text",
              "value": "headline one",
            },
            Object {
              "children": Array [],
              "keyword": "DEADLINE",
              "parent": [Circular],
              "timestamp": "<2018-01-01 Mon>",
              "type": "planning",
            },
            Object {
              "children": Array [],
              "name": "PROPERTIES",
              "parent": [Circular],
              "type": "drawer",
              "value": "key: value
key: value",
            },
          ],
          "keyword": "NEXT",
          "level": 1,
          "parent": [Circular],
          "priority": undefined,
          "tags": Array [],
          "type": "headline",
        },
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "desc": "Here's",
              "parent": [Circular],
              "type": "link",
              "uri": Object {
                "location": "//github.com/xiaoxinghu/orgajs",
                "protocol": "https",
                "raw": "https://github.com/xiaoxinghu/orgajs",
              },
            },
            Object {
              "children": Array [],
              "parent": [Circular],
              "type": "text",
              "value": " to the ",
            },
            Object {
              "children": Array [
                Object {
                  "children": Array [],
                  "parent": [Circular],
                  "type": "text",
                  "value": "crazy",
                },
              ],
              "parent": [Circular],
              "type": "bold",
            },
            Object {
              "children": Array [],
              "parent": [Circular],
              "type": "text",
              "value": " ones, the ",
            },
            Object {
              "children": Array [
                Object {
                  "children": Array [],
                  "parent": [Circular],
                  "type": "text",
                  "value": "misfits",
                },
              ],
              "parent": [Circular],
              "type": "italic",
            },
            Object {
              "children": Array [],
              "parent": [Circular],
              "type": "text",
              "value": ", the ",
            },
            Object {
              "children": Array [
                Object {
                  "children": Array [],
                  "parent": [Circular],
                  "type": "text",
                  "value": "rebels",
                },
              ],
              "parent": [Circular],
              "type": "underline",
            },
            Object {
              "children": Array [],
              "parent": [Circular],
              "type": "text",
              "value": ", the ",
            },
            Object {
              "children": Array [
                Object {
                  "children": Array [],
                  "parent": [Circular],
                  "type": "text",
                  "value": "troublemakers",
                },
              ],
              "parent": [Circular],
              "type": "code",
            },
            Object {
              "children": Array [],
              "parent": [Circular],
              "type": "text",
              "value": ",",
            },
            Object {
              "children": Array [],
              "parent": [Circular],
              "type": "text",
              "value": "the round pegs in the ",
            },
            Object {
              "children": Array [
                Object {
                  "children": Array [],
                  "parent": [Circular],
                  "type": "text",
                  "value": "round",
                },
              ],
              "parent": [Circular],
              "type": "strikeThrough",
            },
            Object {
              "children": Array [],
              "parent": [Circular],
              "type": "text",
              "value": " square holes...",
            },
          ],
          "parent": [Circular],
          "type": "paragraph",
        },
      ],
      "level": 1,
      "parent": [Circular],
      "type": "section",
    },
  ],
  "meta": Object {
    "date": "2018-01-01",
    "title": "hello world",
    "todos": Array [
      "TODO",
      "NEXT",
      "DONE",
    ],
  },
  "type": "root",
}
`;
